### PR TITLE
[fix][cursor] Fix enable maxPosition cursor will read in deap loop

### PR DESCRIPTION
**Pull request by congbobo184**
### Motivation
Now we don't check `OpAddEntry` can read entry is smaller than maxReadPosition, if readPosition is bigger than maxReadPosition, we only can read an empty list as result. In fact, when maxReadPosition have not been updated, we should put read op in to wait state and when maxReadPosition updated, we can notify it and read continue.

### Modifications
1. Add a field named maxReadPosition in`ManagedLedgerImpl` and if read op enable maxReadPosition, we should check this readPosition is bigger than maxReadPosition, if bigger we should put the op into wait state, if smaller or equals we can  do this op.
2. Add a field named `waitingCursorsByMaxReadPosition` in `ManagedLedgerImpl`, when cursor has read op in wait state, we can put this read op in to this queue. If maxReadPosition updated, we will pool it and notify this this read op.
3. In topicTransaction buffer, when any updated maxReadposition op we should sync it to `ManagedLedgerImpl` maxReadPosition.
### Verifying this change
add the test

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API: (no)
  - The schema: (no)
  - The default values of configurations: (no)
  - The wire protocol: (no)
  - The rest endpoints: (no)
  - The admin cli options: (no)
  - Anything that affects deployment: (no)

### Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)
  - If a feature is not applicable for documentation, explain why?
  - If a feature is not documented yet in this PR, please create a followup issue for adding the documentation

### CODE REVIEW

**Comment by eolivelli** [comment]
This change is a big refactor in ManagedCursor, it is not a transaction only related patch.

Please change the title, otherwise it is possible that this change would not be noticed by people not interested in Transactions.

Also I would say that we cannot cherry-pick this change to old branches, as it is adding breaking changes in the internal interfaces (I am pretty sure that some Protocol Handler won't compile)

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java [hunk 1]
```diff
@@ -848,11 +944,14 @@ public boolean cancelPendingReadRequest() {
         if (log.isDebugEnabled()) {
             log.debug("[{}] [{}] Cancel pending read request", ledger.getName(), name);
         }
-        return WAITING_READ_OP_UPDATER.getAndSet(this, null) != null;
+        // cancel all read op
+        return WAITING_READ_OP_UPDATER.getAndSet(this, null) != null
+                | WAITING_READ_OP_UPDATER_BY_MAX_READ_POSITION.getAndSet(this, null) != null;
```
[the remark this answers is not in the document]
Comment (congbobo184) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java:949]: should use "|", we should cancle all wait read op not use "||"

[the remark this answers is not in the document]
Comment (eolivelli) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java:949]: makes sense.
I wonder if we are testing the two branches.
it is possible that in the future someone things that this is a typo and we should not prevent regressions.
do you think that a test will fail if we change this operator to "||" ?
I expect something to fail, if this is not a case I suggest to add a test case

**Comment by TakaHiR07** [comment]
I am wondering why this pr do not invoke updateMaxReadPosition in TopicTransactionBuffer#handleTransactionMessage() ?

**Comment by poorbarcode** [comment]
Since we will start the RC version of `3.0.0` on `2023-04-11`, I will change the label/milestone of PR who have not been merged.
- The PR of type `feature` is deferred to `3.1.0`
- The PR of type `fix` is deferred to `3.0.1`

So drag this PR to `3.0.1`

**Comment by dao-jun** [comment]
It looks the issue fixed by https://github.com/apache/pulsar/pull/14667, closing... feel free to reopen if this PR still ongoing

**Comment by TakaHiR07** [comment]
> It looks the issue fixed by #14667, closing... feel free to reopen if this PR still ongoing

@dao-jun This issue is not fix. I test again in version-3.0.5 and this issue reproduce. This pr still need to be merged.

**Pull request closed without merging**